# Hands-on example

In [1]:
import os
import warnings
import sys

import pandas as pd
import numpy as np

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import ElasticNet

import mlflow
import mlflow.sklearn

## 0. The data

* The data set used in this example is from http://archive.ics.uci.edu/ml/datasets/Wine+Quality
* P. Cortez, A. Cerdeira, F. Almeida, T. Matos and J. Reis.
* Modeling wine preferences by data mining from physicochemical properties. In Decision Support Systems, Elsevier, 47(4):547-553, 2009.

In [2]:
data_path = "data/wine-quality.csv"
data = pd.read_csv(data_path)

data.sample(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
2891            6.9              0.20         0.30            4.70      0.041   
4118            6.5              0.43         0.31            3.60      0.046   
1640            7.8              0.49         0.49            7.00      0.043   
757             6.8              0.22         0.37           15.20      0.051   
3720            6.3              0.34         0.27            2.50      0.024   
3244            6.7              0.21         0.36            8.55      0.020   
2352            6.1              0.28         0.16            1.30      0.060   
3641            6.0              0.25         0.40            5.70      0.052   
1787            7.2              0.26         0.30            2.10      0.033   
1113            7.4              0.18         0.40            1.60      0.047   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
2891                 40.0                 148.0  0.99320  3.16       0.35   
4118                 19.0                 143.0  0.99022  3.15       0.34   
1640                 29.0                 149.0  0.99520  3.21       0.33   
757                  68.0                 178.0  0.99935  3.40       0.85   
3720                 40.0                 152.0  0.99095  3.35       0.60   
3244                 20.0                  86.0  0.99146  3.19       0.22   
2352                 36.0                 126.0  0.99353  3.13       0.46   
3641                 56.0                 152.0  0.99398  3.16       0.88   
1787                 50.0                 158.0  0.99090  3.33       0.43   
1113                 22.0                 102.0  0.99370  3.28       0.44   

      alcohol  quality  
2891     10.2        6  
4118     12.0        8  
1640     10.0        5  
757       9.3        6  
3720     11.9        7  
3244     13.4        7  
2352      8.7        6  
3641     10.5        6  
1787     12.1        7  
1113     10.7        5

## 1. Tracking experiments

### Tracking stores
MLflow supports two types of backend stores: *file store* and *database-backed* store.

- Local file path (specified as file:/my/local/dir), where data is just directly stored locally. Defaults to `mlruns/`
- Database encoded as <dialect>+<driver>://<username>:<password>@<host>:<port>/<database>. Mlflow supports the dialects mysql, mssql, sqlite, and postgresql. For more details, see SQLAlchemy database uri.
- HTTP server (specified as https://my-server:5000), which is a server hosting an MLFlow tracking server.
- Databricks workspace (specified as databricks or as databricks://<profileName>, a Databricks CLI profile.
    
### Artifact stores
- Amazon S3
- Azure Blob Storage
- Google Cloud Storage
- FTP server
- SFTP Server
- NFS
- HDFS

Start the MLflow tracking server by 

```
mlflow server \
    --backend-store-uri /mnt/persistent-disk \
    --default-artifact-root s3://my-mlflow-bucket/ \
    --host 0.0.0.0
    --port 5000
```

or use the default storage method to write to `mlruns/`.

In [3]:
# mlflow server --backend-store-uri mlruns/ --default-artifact-root mlruns/ --host 0.0.0.0 --port 5000
remote_server_uri = "http://127.0.0.1:5000" # set to your server URI
mlflow.set_tracking_uri(remote_server_uri)  # or set the MLFLOW_TRACKING_URI in the env

In [4]:
mlflow.tracking.get_tracking_uri()

'http://127.0.0.1:5000'

In [5]:
exp_name = "ElasticNet_wine_TA"
mlflow.set_experiment(exp_name)

<Experiment: artifact_location='/home/ayanava/mlflow/mlruns/952775695246987617', creation_time=1733344558875, experiment_id='952775695246987617', last_update_time=1733344558875, lifecycle_stage='active', name='ElasticNet_wine_TA', tags={}>

### What do we track?

- **Code Version**: Git commit hash used for the run (if it was run from an MLflow Project)
- **Start & End Time**: Start and end time of the run
- **Source**: what code run?
- **Parameters**: Key-value input parameters.
- **Metrics**: Key-value metrics, where the value is numeric (can be updated over the run)
- **Artifacts**: Output files in any format.

In [6]:
def eval_metrics(actual, pred):
    # compute relevant metrics
    rmse = np.sqrt(mean_squared_error(actual, pred))
    mae = mean_absolute_error(actual, pred)
    r2 = r2_score(actual, pred)
    return rmse, mae, r2


def load_data(data_path):
    data = pd.read_csv(data_path)

    # Split the data into training and test sets. (0.75, 0.25) split.
    train, test = train_test_split(data)

    # The predicted column is "quality" which is a scalar from [3, 9]
    train_x = train.drop(["quality"], axis=1)
    test_x = test.drop(["quality"], axis=1)
    train_y = train[["quality"]]
    test_y = test[["quality"]]
    return train_x, train_y, test_x, test_y


def train(alpha=0.5, l1_ratio=0.5):
    # train a model with given parameters
    warnings.filterwarnings("ignore")
    np.random.seed(40)

    # Read the wine-quality csv file (make sure you're running this from the root of MLflow!)
    data_path = "data/wine-quality.csv"
    train_x, train_y, test_x, test_y = load_data(data_path)

    # Useful for multiple runs (only doing one run in this sample notebook)    
    with mlflow.start_run():
        # Execute ElasticNet
        lr = ElasticNet(alpha=alpha, l1_ratio=l1_ratio, random_state=42)
        lr.fit(train_x, train_y)

        # Evaluate Metrics
        predicted_qualities = lr.predict(test_x)
        (rmse, mae, r2) = eval_metrics(test_y, predicted_qualities)

        # Print out metrics
        print("Elasticnet model (alpha=%f, l1_ratio=%f):" % (alpha, l1_ratio))
        print("  RMSE: %s" % rmse)
        print("  MAE: %s" % mae)
        print("  R2: %s" % r2)

        # Log parameter, metrics, and model to MLflow
        mlflow.log_param(key="alpha", value=alpha)
        mlflow.log_param(key="l1_ratio", value=l1_ratio)
        mlflow.log_metric(key="rmse", value=rmse)
        mlflow.log_metrics({"mae": mae, "r2": r2})
        mlflow.log_artifact(data_path)
        print("Save to: {}".format(mlflow.get_artifact_uri()))
        
        mlflow.sklearn.log_model(lr, "model")

In [7]:
# could also do
# with mlflow.start_run():
#     for epoch in range(0, 3):
#          mlflow.log_metric(key="quality", value=2*epoch, step=epoch)

In [8]:
train(0.5, 0.5)

Elasticnet model (alpha=0.500000, l1_ratio=0.500000):
  RMSE: 0.82224284975954
  MAE: 0.6278761410160693
  R2: 0.12678721972772677
Save to: /home/ayanava/mlflow/mlruns/952775695246987617/e17f882ba2eb43e39ad095e2f1ca6ec4/artifacts


2024/12/05 02:26:05 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


🏃 View run likeable-hawk-1000 at: http://127.0.0.1:5000/#/experiments/952775695246987617/runs/e17f882ba2eb43e39ad095e2f1ca6ec4
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/952775695246987617


In [9]:
train(0.2, 0.2)

Elasticnet model (alpha=0.200000, l1_ratio=0.200000):
  RMSE: 0.7859129997062341
  MAE: 0.6155290394093895
  R2: 0.20224631822892103
Save to: /home/ayanava/mlflow/mlruns/952775695246987617/a7ebc4fb9d4b4078bdb3e4337b02bf4c/artifacts


2024/12/05 02:26:09 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


🏃 View run fortunate-newt-422 at: http://127.0.0.1:5000/#/experiments/952775695246987617/runs/a7ebc4fb9d4b4078bdb3e4337b02bf4c
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/952775695246987617


In [10]:
train(0.1, 0.1)

Elasticnet model (alpha=0.100000, l1_ratio=0.100000):
  RMSE: 0.7792546522251949
  MAE: 0.6112547988118587
  R2: 0.2157063843066196
Save to: /home/ayanava/mlflow/mlruns/952775695246987617/477dec1ecef04d1cb9abfd61749028a0/artifacts


2024/12/05 02:26:11 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


🏃 View run flawless-calf-549 at: http://127.0.0.1:5000/#/experiments/952775695246987617/runs/477dec1ecef04d1cb9abfd61749028a0
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/952775695246987617


### 1.1 Comparing runs
Run `mlflow ui` in a terminal or `http://your-tracking-server-host:5000` to view the experiment log and visualize and compare different runs and experiments. The logs and the model artifacts are saved in the `mlruns` directory (or where you specified).

## 2. Packaging the experiment as a MLflow project as conda env

Specify the entrypoint for this project by creating a `MLproject` file and adding an conda environment with a `conda.yaml`. You can copy the yaml file from the experiment logs.

To run this project, invoke `mlflow run . -P alpha=0.42`. After running this command, MLflow runs your training code in a new Conda environment with the dependencies specified in `conda.yaml`.


## 3. Deploy the model

Deploy the model locally by running 

`mlflow models serve -m mlruns/0/f5f7c052ddc5469a852aa52c14cabdf1/artifacts/model/ -h 0.0.0.0 -p 1234`

Test the endpoint:

`curl -X POST -H "Content-Type:application/json; format=pandas-split" --data '{"columns":["alcohol", "chlorides", "citric acid", "density", "fixed acidity", "free sulfur dioxide", "pH", "residual sugar", "sulphates", "total sulfur dioxide", "volatile acidity"],"data":[[12.8, 0.029, 0.48, 0.98, 6.2, 29, 3.33, 1.2, 0.39, 75, 0.66]]}' http://0.0.0.0:1234/invocations`

You can also simply build a docker image from your model

`mlflow models build-docker -m mlruns/1/d671f37a9c7f478989e67eb4ff4d1dac/artifacts/model/ -n elastic_net_wine`

and run the container with

`docker run -p 8080:8080 elastic_net_wine`.

Or you can directly deploy to AWS sagemaker or Microsoft Azure ML.

## 4. Tagging runs

In [15]:
from datetime import datetime
from mlflow.tracking import MlflowClient

client = MlflowClient()
experiments = client.search_experiments() # returns a list of mlflow.entities.Experiment
print(experiments)

[<Experiment: artifact_location='/home/ayanava/mlflow/mlruns/952775695246987617', creation_time=1733344558875, experiment_id='952775695246987617', last_update_time=1733344558875, lifecycle_stage='active', name='ElasticNet_wine_TA', tags={}>, <Experiment: artifact_location='/home/ayanava/mlflow/mlruns/0', creation_time=1733344529844, experiment_id='0', last_update_time=1733344529844, lifecycle_stage='active', name='Default', tags={}>]


In [16]:
# get the run
_run = client.get_run(run_id="477dec1ecef04d1cb9abfd61749028a0")
print(_run)

<Run: data=<RunData: metrics={'mae': 0.6112547988118587,
 'r2': 0.2157063843066196,
 'rmse': 0.7792546522251949}, params={'alpha': '0.1', 'l1_ratio': '0.1'}, tags={'mlflow.log-model.history': '[{"run_id": "477dec1ecef04d1cb9abfd61749028a0", '
                             '"artifact_path": "model", "utc_time_created": '
                             '"2024-12-04 20:56:09.543939", "model_uuid": '
                             '"9a077ffef52b41fa95e4db3c0cab3b66", "flavors": '
                             '{"python_function": {"model_path": "model.pkl", '
                             '"predict_fn": "predict", "loader_module": '
                             '"mlflow.sklearn", "python_version": "3.12.4", '
                             '"env": {"conda": "conda.yaml", "virtualenv": '
                             '"python_env.yaml"}}, "sklearn": '
                             '{"pickled_model": "model.pkl", '
                             '"sklearn_version": "1.5.2", '
                            

In [17]:
# add a tag to the run
dt = datetime.now().strftime("%d-%m-%Y (%H:%M:%S.%f)")
client.set_tag(_run.info.run_id, "deployed", dt)

# Connect to a postgesql db

Connect to the db:
    
`sudo -u postgres psql`

Create a user and a database for the tracking server:
    
```
CREATE DATABASE mlflow;
CREATE USER mlflow WITH ENCRYPTED PASSWORD 'mlflow';
GRANT ALL PRIVILEGES ON DATABASE mlflow TO mlflow;
```

```
mlflow server --backend-store-uri postgresql://mlflow:mlflow@localhost/mlflow --default-artifact-root file:/home/tobias/Projects/mlflow_talk/pydataberlin-2019/mlruns/ -h 0.0.0.0 -p 8000
```

Run the notebook again with this tracking server.

Look at the db:

`psql mlflow`

`SELECT * FROM experiments;`

`SELECT * FROM runs;`